src/backend/distributed/commands/multi_copy.c

```diff
@@ -1195,6 +1233,195 @@ AppendCopyBinaryFooters(CopyOutState footerOutputState)
 }
 
 
+/*
+ * CopyToAppendPartitionedTable implements the COPY table_name FROM ... for
+ * append-partitioned tables.
+ */
+static void
+CopyToAppendPartitionedTable(CopyStmt *copyStatement, char *completionTag)
+{
+	Oid relationId = RangeVarGetRelid(copyStatement->relation, NoLock, false);
+	FmgrInfo *columnOutputFunctions = NULL;
+
+	/* allocate column values and nulls arrays */
+	Relation distributedRelation = heap_open(relationId, RowExclusiveLock);
+	TupleDesc tupleDescriptor = RelationGetDescr(distributedRelation);
+	uint32 columnCount = tupleDescriptor->natts;
+	Datum *columnValues = palloc0(columnCount * sizeof(Datum));
+	bool *columnNulls = palloc0(columnCount * sizeof(bool));
+
+	EState *executorState = CreateExecutorState();
+	MemoryContext executorTupleContext = GetPerTupleMemoryContext(executorState);
+	ExprContext *executorExpressionContext = GetPerTupleExprContext(executorState);
+
+	ShardConnections *shardConnections =
```

Comment:
Where we assign the shard connections hash in CopyToHashOrRangePartitionedTable there's a note about PG_TRY subtleties because that variable is used in the PG_CATCH (which wouldn't work if it was assigned within PG_TRY), it might be useful to add the same note here or refer to the other one.